### PR TITLE
UPP thread safe 

### LINKED ISSUE #270: Fix thread violation issues in UPP
A decision was made 10 years ago to focus on improving UPP's efficiency using MPI.
As a consequence, UPP has not been thread safe for at least a decade and users have
been advised to run with OMP_NUM_THREADS=1

UPP recently switched to use cmake as request from EIB.
It was quickly discovered the use of cmake in UPP compilation resulted in intrinsic use of OpenMP
which caused UPP to generate different output when using different numbers of threads both inline
and stand-alone.

Qingfu found the following fields to be different when running with different numbers of threads in inline post:
821:2269498972:vt=2019070112:4000 m above ground:12 hour fcst:REFD Reflectivity [dB]:
 822:2270643799:vt=2019070112:1000 m above ground:12 hour fcst:REFD Reflectivity [dB]:
1005:2836535014:vt=2019070112:120-90 mb above ground:12 hour fcst:TMP Temperature [K]:
1009:2852643053:vt=2019070112:120-90 mb above ground:12 hour fcst:VGRD V-Component of Wind [m/s]:
1020:2896927589:vt=2019070112:surface:12 hour fcst:4LFTX Best (4 layer) Lifted Index [K]:
 1022:2901122654:vt=2019070112:180-0 mb above ground:12 hour fcst:CIN Convective Inhibition [J/kg]:

Jesse found only Radar relativities are different in his stand-alone tests using different threads.

Qingfu provided a fix which Jesse verified fixing radar ref irreproducibility in his stand-alone tests.

### PR DESCRIPTION
**Pull request by JesseMeng-NOAA**
Improvement for threading safe

### CODE REVIEW

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Could you provide the information of tests you conducted in this PR's description session?

**Review by JamesAbeles-NOAA — APPROVED** [review]
It seems as if Jesse has tested the code using multiple thread counts and the answers are indentical. I approve the changes but would like some additional testing by others to make sure there are no other issues.

File: ush/gfs_nceppost.sh [hunk 1]
```diff
@@ -274,6 +274,7 @@ fi
 
 # allow threads to use threading in Jim's sp lib
 # but set default to 1 
+export OMP_NUM_THREADS=$threads
```
Comment (WenMeng-NOAA) [inline ush/gfs_nceppost.sh]: @JesseMeng-NOAA This change might have impact on the config. of the upstream global workflow. Please remove it.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 2]
```diff
@@ -843,18 +843,17 @@ PROGRAM WRFPOST
...
         ENDIF
 !
 !       IF GRIB2 read out post aviable fields xml file and post control file
 !
         if(grib == "grib2") then
-          btim=timef()
+          btim=mpi_wtime()
           call READ_xml()
-          READxml_tim = READxml_tim + (timef() - btim)
+          READxml_tim = READxml_tim + (mpi_wtime() - btim)
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f:858]: @JesseMeng-NOAA Please update the timers in other routines e.g. PROCESS.f with Jim's changes.

**Comment by JesseMeng-NOAA** [comment]
The reason for adding the following line in gfs_nceppost.sh,

+export OMP_NUM_THREADS=$threads

right before calling the executable is because  OMP_NUM_THREADS  will be
reset to 1 by "module purge" and "module load EnvVars/1.0.2"

And in all the run scripts the variable "threads" and "OMP_NUM_THREADS" are
specified as

export threads=1
export OMP_NUM_THREADS=$threads

I am curious about the usage of "threads" instead of just assigning the
OMP_NUM_THREADS value directly.
If we do not want the line "export OMP_NUM_THREADS=$threads" in
gfs_nceppost.sh, we could modify all the regression test run scripts to set
the  OMP_NUM_THREADS value after
module load.

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Can you set thread in your driver scripts?

**Comment by JesseMeng-NOAA** [comment]
Yes we can. And all the regression test run scripts need to be modified to
set the OMP_NUM_THREADS value after
module load.
==============================================
Dr. Jesse Meng
IMSG at NOAA/NWS/NCEP/EMC
5830 University Research Ct. Room 2037
College Park, MD 20740
email: ***@***.***
==============================================

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA The OMP_NUM_THREADS would be a variable from the global workflow. You may add changes in your local tests.

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Got it. We could update the job cards in UPP regression tests after threading is turned on in UPP.

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA You might adapt some of UPP regression test scripts for your own driver scripts. It would be easy for setting up  various configs. in your tests.

**Comment by JesseMeng-NOAA** [comment]
Jim modified TIMEF.f with
!        timef=etime(et)
!        timef=timef*1.e3
        timef=mpi_wtime()

So using timef() in PROCESS.f is still good.

File: ush/gfs_nceppost.sh (hunk 1 above)
Comment (JesseMeng-NOAA) [reply ush/gfs_nceppost.sh]: Done!

File: sorc/ncep_post.fd/WRFPOST.f [hunk 3]
```diff
@@ -174,7 +174,7 @@ PROGRAM WRFPOST
       integer      :: kpo,kth,kpv
       real,dimension(komax) :: po,th,pv
       namelist/nampgb/kpo,po,kth,th,kpv,pv,fileNameAER,d3d_on,gocart_on,popascal &
-                     ,hyb_sigp,rdaod
+                     ,hyb_sigp
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @JesseMeng-NOAA Please add "rdaod" back.

File: sorc/ncep_post.fd/WRFPOST.f [hunk 4]
```diff
@@ -307,7 +307,6 @@ PROGRAM WRFPOST
         gocart_on   = .false.
         popascal    = .false.
         fileNameAER = ''
-        rdaod       = .false.
```
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/WRFPOST.f]: @JesseMeng-NOAA Add this line back.

Comment (JesseMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: Done

File: sorc/ncep_post.fd/WRFPOST.f (hunk 3 above)
Comment (JesseMeng-NOAA) [reply sorc/ncep_post.fd/WRFPOST.f]: Done

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA There are few sub process routines not having timers in  PROCESS.f. Would you consider to add? For example, 
 MDL2THANDPV, CALRAD_WCLOUD, MDL2STD_P.

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA There are few sub process routines not having timers in PROCESS.f. Would you consider to add? For example,
> MDL2THANDPV, CALRAD_WCLOUD, MDL2STD_P.

Done!

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Could you update VERSION with 10.0.5?

**Comment by JesseMeng-NOAA** [comment]
> @JesseMeng-NOAA Could you update VERSION with 10.0.5?

Build a upp/10.0.5 ?
Please provide the procedure. Thanks!

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA That's right. You update the file VERSION, 
change
   10.0.4
into 
    10.0.5

**Comment by WenMeng-NOAA** [comment]
@JesseMeng-NOAA Could you get the file VERSION under EMC_post/ updated?

**Comment by WenMeng-NOAA** [comment]
The UPP standalone tests were conducted on WCOSS-Dell, WCOSS-Cray and Hera. No changed results were found. This PR is ready for merging.

**Review by WenMeng-NOAA — APPROVED** [review]

**Pull request merged**